torch_geometric/nn/pool/approx_knn.py

```diff
@@ -0,0 +1,55 @@
+# This file has code to create a KNN graph with approximate
+# nearest neighbors. We can call it approx-knn.
+
+import torch
+from pynndescent import NNDescent
+
+
+def approx_nn(x, y, k, batch_x=None, batch_y=None):
+    if batch_x is None:
+        batch_x = x.new_zeros(x.size(0), dtype=torch.long)
+
+    if batch_y is None:
+        batch_y = y.new_zeros(y.size(0), dtype=torch.long)
+
+    x = x.view(-1, 1) if x.dim() == 1 else x
+    y = y.view(-1, 1) if y.dim() == 1 else y
+
+    assert x.dim() == 2 and batch_x.dim() == 1
+    assert y.dim() == 2 and batch_y.dim() == 1
+    assert x.size(1) == y.size(1)
+    assert x.size(0) == batch_x.size(0)
+    assert y.size(0) == batch_y.size(0)
+
+    min_xy = min(x.min().item(), y.min().item())
+    x, y = x - min_xy, y - min_xy
+
+    max_xy = max(x.max().item(), y.max().item())
+    x, y, = x / max_xy, y / max_xy
+
+    # Concat batch/features to ensure no cross-links between examples exist.
+    x = torch.cat([x, 2 * x.size(1) * batch_x.view(-1, 1).to(x.dtype)], dim=-1)
+    y = torch.cat([y, 2 * y.size(1) * batch_y.view(-1, 1).to(y.dtype)], dim=-1)
+
+    # Default metric is euclidean
+    # More can be found in the
+    # documentation here - https://pynndescent.readthedocs.io/en/latest/
```

Comment:
I think this scheme here only works for Euclidean due to how we integrate `batch` into `x`/`y`.

Reply:
ok, can you elaborate?